webtau-core/src/main/java/com/twosigma/webtau/expectation/equality/handlers/SetCompareToHandler.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twosigma.webtau.expectation.equality.handlers;
+
+import com.twosigma.webtau.expectation.ActualPath;
+import com.twosigma.webtau.expectation.equality.CompareToComparator;
+import com.twosigma.webtau.expectation.equality.CompareToHandler;
+import com.twosigma.webtau.expectation.equality.CompareToResult;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+public class SetCompareToHandler implements CompareToHandler {
+    @Override
+    public boolean handleEquality(Object actual, Object expected) {
+        return actual instanceof Set && expected instanceof Set;
+    }
+
+    @Override
+    public void compareEqualOnly(CompareToComparator compareToComparator, ActualPath actualPath, Object actual, Object expected) {
+        Set<?> actualSet = (Set) actual;
+        Set<?> expectedSet = (Set) expected;
+
+        Comparator comparator = new Comparator(compareToComparator, actualPath, actualSet, expectedSet);
+        comparator.compare();
+    }
+
+    private class Comparator {
+        private final CompareToComparator comparator;
+        private final CompareToComparator localComparator;
+        private final ActualPath actualPath;
+        List<?> actualLeft;
+        List<?> expectedLeft;
+
+        Comparator(CompareToComparator compareToComparator, ActualPath actualPath, Set<?> actual, Set<?> expected) {
+            this.comparator = compareToComparator;
+            this.localComparator = CompareToComparator.comparator(compareToComparator.getAssertionMode());
+            this.actualPath = actualPath;
+            this.actualLeft = new ArrayList<>(actual);
+            this.expectedLeft = new ArrayList<>(expected);
+        }
+
+        void compare() {
+            expectedLeft.removeIf(this::checkActualAndRemoveIfMatch);
```

Comment:
It's slightly confusing having the `removeIf` use a predicate with the word `removeIf` in it.  If I understand this correct, one removes from expected and the other from actual?  If that's the case, would a better name for the predicate be `checkAndRemoveFromActualIfMatch` or something along those lines?  Basically I think it would help to make it obvious from where the removal is happening in the predicate.

An alternative approach would be to do something like:
```
Map<Object, Object> matches = expectedLeft
                    .stream()
                    .map(this::getMatches)
                    .filter(Objects::nonNull)
                    .collect(Collectors.toMap(
                            Pair::getKey,
                            Pair::getValue
                    ));
            
            expectedLeft.removeAll(matches.keySet());
            actualLeft.removeAll(matches.values());
```
The `getMatches` is basically identical to `checkActualAndRemoveIfMatch` except it doesn't remove and instead of `true` returns `Pair.of(expected, actual)` and `null` instead of `false`.  That's probably not the best way to write that code but hopefully you get what I mean.  It's less efficient of course but arguably easier to understand.

Reply:
I like `getMatches`, definitely more functional. I do not like maps returning empty and then filtering out bit though. 
I will proceed with rename since performance in case of sets is already a small concern.

Reply:
Yeah, the empty filtration bothered me a bit too.